qutebrowser/browser/downloads.py

```diff
@@ -894,110 +894,212 @@ def _on_data_changed(self, idx, *, webengine):
+        except ValueError:
+            raise cmdexc.CommandError(
+                    "Invalid index set '{}'!".format(indexset)
+                  )
+
+        # identify downloads that match the criterion, as well as those that
+        # don't exist
+        matchs = []
+        nomatchs = []
+        noexists = []
+        for i in indexset_p:
+            j = i-1 # j is the download item when counting from 0, i is the one
+                    # when counting from 1 - do we need this? maybe one day we
+                    # should count downloads, everywhere, from 0, and get away
+                    # with this?
+            try:
+                d = self[j]
+                if criterion(d):
+                    matchs.append(d)
+                else:
+                    nomatchs.append(i)
+            except IndexError:
+                noexists.append(i)
+
+        return (matchs, nomatchs, noexists)
 
     @cmdutils.register(instance='download-model', scope='window')
     @cmdutils.argument('count', count=True)
-    def download_cancel(self, all_=False, count=0):
+    def download_cancel(self, indexset=None, all_=False, count=0):
         """Cancel the last/[count]th download.
 
         Args:
-            all_: Cancel all running downloads
+            indexset: Specify the set of download indexes to delete. E.g.
+                      '2,5,6-last,3', where 'last' is special keyword that
+                      denotes that latest download index. If you specify
+                      numbers smaller than 1, or greater than 'last', then
+                      those numbers will be silently ignored.
+            all_: Cancels all downloads. This is a shortcut to using the list
+                  specification '1-last'.
             count: The index of the download to cancel.
         """
-        downloads = self._all_downloads()
-        if all_:
-            for download in downloads:
-                if not download.done:
-                    download.cancel()
-        else:
-            try:
-                download = downloads[count - 1]
-            except IndexError:
-                self._raise_no_download(count)
-            if download.done:
-                if not count:
-                    count = len(self)
-                raise cmdexc.CommandError("Download {} is already done!"
-                                        .format(count))
-            download.cancel()
+
+        # find matched downloads
+        (matchs, nomatchs, noexists) = self._select_downloads(
+                                           count,
+                                           all_,
+                                           indexset,
+                                           lambda d: not d.done
+                                       )
+
+        # cancel them
+        for d in matchs:
+            d.cancel()
+
+        # show errors if needed
+        if len(nomatchs):
+            message.error("Downloads {} are already done!".format(
```

Comment:
What would you need to edit in `_select_downloads` when `download_pause` is added, with my suggestion?

The thing is, if we add some logic to show singular/plural correctly when there's only one download, it'd be much nicer to only have that logic once instead of everywhere where it's called. Also, there's only one clear return value in that case. I think the benefits are more than the drawbacks (showing errors in `_select_downloads`).